### PR TITLE
Make databricks sql hook return a namedtuple

### PR DESCRIPTION
**Pull request by Joffreybvn**
This is a fix proposal for https://github.com/apache/airflow/pull/36161#pullrequestreview-1779199804

The [ADR for Airflow's DB API](https://github.com/apache/airflow/blob/main/airflow/providers/common/sql/doc/adr/0002-return-common-data-structure-from-dbapihook-derived-hooks.md) specifies that DbApiHooks needs to return a tuple / list of tuples. The databricks hook by defaults returns a dict-like [`databricks.Row`](https://github.com/databricks/databricks-sql-python/blob/62eb1d407ab03fc2c57a6626bec2173dbdd8bbb3/src/databricks/sql/types.py#L24) object which we now transform into a new nametuple - _which is compliant with the 'common data structure' specification._

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by potiuk — APPROVED** [review]
Looks good

**Comment by potiuk** [comment]
@Joffreybvn -> do you have more things to add to it or you think it's ready to review ?

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (utkarsharma2) [inline airflow/providers/databricks/hooks/databricks_sql.py]: The same testcase from earlier PR would still fail here, right? 

```
statement = f"DESCRIBE TABLE {table.name};"
hook.run(statement, parameters=parameters, handler=lambda x: x.fetchall())
```
with this code hook.run() will return 
[SerializableRow(),SerializableRow(), SerializableRow()] instead of [row(), row(), row()]

Since `Row` is not the same as `namedtuple` class this is still a breaking change.

Comment (phanikumv) [reply airflow/providers/databricks/hooks/databricks_sql.py]: Agree with @utkarsharma2  comment, we still need to update the code in this PR.

**Review by phanikumv — CHANGES_REQUESTED** [review]
We need to address https://github.com/apache/airflow/pull/36205#discussion_r1425309132

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (potiuk) [reply airflow/providers/databricks/hooks/databricks_sql.py]: yep

Comment (Joffreybvn) [reply airflow/providers/databricks/hooks/databricks_sql.py]: Indeed, this PR is not about preventing the breaking change. But a proposal to break less workflows - _the simple tuple/list is too agressive_ - and reach the goal of [this ADR](https://github.com/apache/airflow/blob/main/airflow/providers/common/sql/doc/adr/0002-return-common-data-structure-from-dbapihook-derived-hooks.md).

The databricks.sql.Row object is a subclass of tuple, which implements:
- a dict-like interface -> breaking change - not available in a namedtuple
- a 'asDict()' method -> breaking change - but namedtuple implements ['_asdict()'](https://docs.python.org/3/library/collections.html#collections.somenamedtuple._asdict)
- fields accessible by name -> supported by namedtuple

Thus users of the dict interface have to rewrite their code. With a major version bump of the provider, would you agree on this change ?

Comment (potiuk) [reply airflow/providers/databricks/hooks/databricks_sql.py]: Yes. We should avoid breaking change like that

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (bolkedebruin) [inline airflow/providers/databricks/hooks/databricks_sql.py]: Please remove the JSON remark. It is either XCom serializable, serde serializable or just serializable. JSON has no meaning here.

**Review by bolkedebruin — CHANGES_REQUESTED** [review]
Given all the discussions around this earlier I'm a bit picky on the naming. We are not making something JSON serializable - but according to the earlier discussion we make the SQL hook return the specified NamedTuple format. 

I wonder btw if we should not have a typedef for it if this is part the the DbAPI. That would make things conceptually stronger and less ambiguous. 

If you insist on naming it something serializable please leave out JSON. It has nothing to do with it.

Can we further be clear on what this PR wants to achieve in the commit message? The title and message are now confusing.

My suggestion:

```
Return SerializableRow in databricks sql hook

The ADR for Airflow' s DB API specifies it needs to return a 
named tuple SerialzableRow or a list of them.  The databricks
hook by defaults returns a dict which we now transform to a
SerializableRow.
```

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (potiuk) [reply airflow/providers/databricks/hooks/databricks_sql.py]: > I wonder btw if we should not have a typedef for it if this is part the the DbAPI. That would make things conceptually stronger and less ambiguous.

Yes. It was missing from the beginning. Agree JSON in the description is wrong.

Comment (Joffreybvn) [reply airflow/providers/databricks/hooks/databricks_sql.py]: Thanks for the re-scoping ! I tried to be reactive on this one, but did not put enough care in the namings. I updated the code and the PR message.

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (Joffreybvn) [reply airflow/providers/databricks/hooks/databricks_sql.py]: I put the serializable namedtuple behind a `return_serializable` flag, which is False by default. With a warning for a future release of the provider, where it would turn to True by default. And eventually later the flag can disappear too to only return the namedtuple.

What do you think ?

**Review by potiuk — APPROVED** [review]
I like the way it is now.

**Comment by potiuk** [comment]
@bolkedebruin @phanikumv -> please take a look and see what you think. I am all for this direction

**Comment by Joffreybvn** [comment]
The databricks provider was already released with the breaking change-> [5.0.1](https://pypi.org/project/apache-airflow-providers-databricks/5.0.1/). Maybe this version should be removed / replaced by a new "5.0.1" one ?

**Comment by potiuk** [comment]
> The databricks provider was already released with the breaking change-> [5.0.1](https://pypi.org/project/apache-airflow-providers-databricks/5.0.1/). Maybe this version should be removed / replaced by a new "5.0.1" one ?

Nope. We can at most yank it. There is no way to replace existing released version. Luckily it has not been yet in constraints of any Airflow version, so we can do it without any serious consequences. Let me actually yank it now.

**Comment by potiuk** [comment]
Yanked

**Comment by potiuk** [comment]
Also cc: @ephraimbuddy -> since we yanked 5.0.1,  I will rerun the last build for v2-8-stable to refresh constraints and move the tag once done (currentlly 5.0.1 was in the constraints planned to release for 2.8.0rc3)

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (bolkedebruin) [inline airflow/providers/databricks/hooks/databricks_sql.py]: Who will set this and why and when? Is this a kind of backwards compatibility? Will this not generate a lot of warnings?

**Comment by potiuk** [comment]
Constraints with databricks 5.0.0 updated https://github.com/apache/airflow/commit/33121932d4cdcd2da27756ac149791365a88f08c and tag for 2.8.0rc3 moved @ephraimbuddy

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (Joffreybvn) [reply airflow/providers/databricks/hooks/databricks_sql.py]: Yes, this is to ensure the change is backward compatible.

Users of the hook will be warned. Users of the `DatabricksSqlOperator` won't, because the operator was already returning a list of tuple - which becomes a list of namedtuple with this PR. Thus, it will generate noise only to hook users.

The goal is to effectively do a breaking change later. But if we don't want that - or not _that_ soon, I can check to print the warning only when the provider loads ?

Comment (Joffreybvn) [reply airflow/providers/databricks/hooks/databricks_sql.py]: Will clarify the docstring and the warning

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (bolkedebruin) [inline airflow/providers/databricks/hooks/databricks_sql.py]: Can we have this type hinted? Why is it not a static method anymore?

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (bolkedebruin) [inline airflow/providers/databricks/hooks/databricks_sql.py]: If something else else a list / row gets in this implicitly returns what was given. Do we want that?

Comment (Joffreybvn) [reply airflow/providers/databricks/hooks/databricks_sql.py]: Not a strong opinion, but I think we do want that:

The `result` input of this function is the output of the [`handler()`](https://github.com/apache/airflow/blob/280b8b30061af2dec60ac83f9fafb92b3f0663c1/airflow/providers/databricks/hooks/databricks_sql.py#L243C58-L243C65) method. This method can be one of the default ones of Airflow (doing simply *fetchone()* or *fetchall()*), but also a totally custom handler injected via the `handler` parameter. Maybe the user want to do something else/something more after fetching, which returns something else than a Row / list[Row] ? Even if it's not the right place to transform the data IMO, blocking anything else than Row/list[Row] can potentially break workflows.

I will improve the isinstance on the list to check if this list contains Row object. And otherwise, the list get ignored too and returned as-is.

Comment (Joffreybvn) [reply airflow/providers/databricks/hooks/databricks_sql.py]: The same is applied to the ODBCHook

**Comment by utkarsharma2** [comment]
@bolkedebruin For a long-term solution would it make sense to have a way for a provider to register a custom serializer/deserializer for objects its operator might produce? We can extend [this logic](https://github.com/apache/airflow/blob/a68b4194fe7201bba0544856b60c7d6724da60b3/airflow/serialization/serde.py#L325) to include providers

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (Joffreybvn) [reply airflow/providers/databricks/hooks/databricks_sql.py]: Not a staticmethod anymore because checking self.return_serializable is needed. I added typings

**Comment by Joffreybvn** [comment]
Update on this PR:
- I removed all mention of serialization / serializable for clarity. Docstring and methods are now speaking about "common data structure".
- The new `_make_common_data_structure` is fully typed, both in databricks and odbc
- If a user have a custom handler that transform the Row object into something else, his workflow won't break

**Comment by potiuk** [comment]
Well, We've just released ODBC **with** make_serializable . However we can still Yank it if we have a quick approval @bolkedebruin @phanikumv ?

**Comment by potiuk** [comment]
One good solution could be to add a deprecated call to _make_serializable. Just to account for odbc already released maybe ?

**Comment by Joffreybvn** [comment]
Why does it needs to be yanked ?

`_make_serializable` is an internal method, hook users are not affected by how it is named. And the only two hooks that implement it are the odbc and databricks ones.

**Comment by potiuk** [comment]
Because it will work differently with just released odbc provider. Unless we call mak_serializable in the future common.sql the odbc provider will behave differently with currently released common.sql and the one that you just modified (to be released in the future)


We have to remember that you can have various common.sql and odbc combinations. We canlikelymake future odbc provider depend on future version of common.sql but we can't restrict the just released odbc to use either current or future common.sql

**Comment by Joffreybvn** [comment]
Added a deprecated ` _make_serializable` in the odbc hook. Databricks doesn't need it because it has a custom run method.

**Review by utkarsharma2 — APPROVED** [review]

File: airflow/providers/databricks/provider.yaml
Comment (potiuk) [inline airflow/providers/databricks/provider.yaml:62]:
```suggestion
  - apache-airflow-providers-common-sql>=1.9.0
```

UPDATE: nope, it's good.

Comment (potiuk) [reply airflow/providers/databricks/provider.yaml:62]: Here and in other places

**Review by phanikumv — APPROVED** [review]

File: airflow/providers/databricks/provider.yaml
Comment (potiuk) [reply airflow/providers/databricks/provider.yaml:62]: Ah no.... scratch that... 1.9.1 is fine (looked at all version of provider.yaml of common.sql). Can you, however add 1.9.1 in the `provider.yaml` as it will not pass the tests. When two providers depends on each other with min-version, the min-version has to be added to provider.yaml of the provider that is depended on.

**Comment by potiuk** [comment]
Hey @bolkedebruin - will you be able to re-review this one as we need to release RC3 soon? I know you are likely away from the keyboard, but I think all of your concerns are addressed now and so when we get it green, I will simply proceed with the release if yoy are not around by then.

File: airflow/providers/odbc/hooks/odbc.py
Comment (potiuk) [inline airflow/providers/odbc/hooks/odbc.py]: Actually what we would have to do, to not yank the current odbc provider is to add code in the common.sql to check if it's there and call it. Adding it here while we also add >= 1.9.1 on the common.sql makes no sense (because 1.9.1 will never call it)

The back-compatibilty here is to let the "just released" odbc provider to work with the new `"common.sql" - not the other way round.

But maybe better - just remove it and I will simply yank the odbc provider we just released, it's going to be way simpler @Joffreybvn

**Comment by bolkedebruin** [comment]
I'll have a look in the next few days.

File: airflow/providers/odbc/hooks/odbc.py
Comment (Joffreybvn) [reply airflow/providers/odbc/hooks/odbc.py]: I think we need this back-compat fix anyway because the `make_serializable´ was introduced in odbc two patches ago (in [4.2.0](https://airflow.apache.org/docs/apache-airflow-providers-odbc/4.3.0/changelog.html#id2)).

Thus, I added it with a warning. Airflow 2.8.1 constraint will need this latest odbc provider version to not annoy users with the warning. So that only people who stick with odbc==4.2.0 or odbc==4.3.0 while upgrading common.sql get warned.

**Review by potiuk — APPROVED** [review]
Very cool! Thanks. 

Yeah. Providers' coupling vs. back-compat makes things quite a bit more difficult 

Thanks for being responsive and keep up with maintainers having different opinions and arguments

**Comment by potiuk** [comment]
cc: @bolkedebruin -  if you  could take a look, that would unblock releasing the RC3 wave of providers.

**Comment by potiuk** [comment]
@bolkedebruin - would love to get this one out :)

**Comment by bolkedebruin** [comment]
checking now

**Comment by bolkedebruin** [comment]
Overall I like the change quite a lot, still I am confused by what is `_make_common _data_structure` is doing in `DBApiHook`. Its comments says that the output is a common format but that format is not stipulated neither checked for or enforced in that function. The implementing functions do a better job, but their signatures are still confusing. Why can they return/accept None? Why do they return T if conversion needs to happen -> you dont apply a common structure then.

can the signature and (pseudo) code of that base function be and others keep using that signature?

```python
T = TypeVar("T")

def _make_common _data_structure(self, result:  T | Sequence[T]) -> tuple | list[tuple]:
  if instance(T,  Sequence):
    return cast(list[tuple], [rows_object(*row) for row in rows])
  
  return cast(tuple, T)
```

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (bolkedebruin) [inline airflow/providers/databricks/hooks/databricks_sql.py]: Potentially you are checking thousands of rows here and `isinstance` is quite slow as it needs to walk the inheritance tree. Is there a way to check a attribute(s) to verify you are dealing with a Row?

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (bolkedebruin) [inline airflow/providers/databricks/hooks/databricks_sql.py]: As explained in the comment `list[T] | T` does not specify a common format. It just takes the input and returns the same. That does not make sense. `list[tuple] | tuple` should be sufficient. And I'd prefer it to be without `None` as that seems not compatible with a 'common_data_structure`

Comment (potiuk) [reply airflow/providers/databricks/hooks/databricks_sql.py]: As explained already in previous discussions I plan to add it in a follow up PR - together with ADR detailing some more of it. I think we should take a few more steps to clarify it and I even have issue defined here and assigned to me: https://github.com/apache/airflow/issues/36224

Perfect is an enemy of good. I suggest we complete this thing now and then add typing as seperate PR.

Is that ok @bolkedebruin  ? Can we do it this way and not circle back the same comment in this PR?

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (potiuk) [reply airflow/providers/databricks/hooks/databricks_sql.py]: What would be your suggestion here @bolkedebruin ? I guess it is not blocking but simple optimisation suggestion? If so - is that fine if we separate it as a follow up issue to optimize it? I guess this is not something that BLOCKS this pr? 

I think we can definitely optimise it in the future, however, again - perfect is the enemy of good and it's holding back release of a new waves of providers . My suggestion is to get it out now and separate it as a follow-up PR.

Is that OK @bolkedebruin ?

**Comment by bolkedebruin** [comment]
Lets leave this to @Joffreybvn if he want to address the issues. I'm not fee entirely happy with the "perfect is the enemy of good" remark. I don't think it applies here. I understand that this has been a tough task because of the earlier discussions and confusion. Nevertheless, I am not sure if we reached the state of "good" yet. 

@Joffreybvn so up to you.

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (bolkedebruin) [reply airflow/providers/databricks/hooks/databricks_sql.py]: an `isinstance` check can take between 15%-60% more time than an attribute check. This change affects current operations, so maybe this is an pre-mature optimization or maybe we do want to apply it. I leave this to @Joffreybvn as he is using it I assume.

**Comment by Joffreybvn** [comment]
The result input of `_make_common _data_structure` is the output of the [handler()](https://github.com/apache/airflow/blob/280b8b30061af2dec60ac83f9fafb92b3f0663c1/airflow/providers/databricks/hooks/databricks_sql.py#L243C58-L243C65) function. Which can be injected by the user, to potentially do more than *fetchone()* or *fetchall()* (e.g. to transform Row into dicts ?).  IMO, user should not transform data there, but `T` and `list[T]` represent the potential, already transformed data structure.

But if we consider `T` and `list[T]` are pyodbc.Row in the ODBC provider, and databricks.sql.Row in Databricks' one, then we don't even need complex `isinstance` checks. And we can indeed easily return tuple/list[tuple] only. That simplify the current implementation, things become less ambiguous - which is great !

But we'll break worflows of users doing bad stuff in the handler. And a [breaking change a bit too aggressive](https://github.com/apache/airflow/pull/36161#pullrequestreview-1779199804) was the root cause of this PR. Totally fine to do the change, as long as we are aware of that.

Thus I'll modify this PR (doing it quickly today),

**Comment by bolkedebruin** [comment]
I'm mindful of being too aggressive, but if we can nudge the use towards a better path I'm all for it.

**Comment by potiuk** [comment]
Well. If we are all ok with breaking change here, I am also fine. The whole `provider's` approach is specifically designed to allow breaking changes in providers. Users can easliy downgrade to previous provider versions and they are generally "future-compatible" because we locked down Hook/Operator API a long time  ago. 

Also `common.sql` is designed to support it (but a bit differently as this is the 'supporting` provider) - from the very beginning we designed and completed `common.sql` to be alwys backwards compatible - because we have many providers depend on it, so we can never make `2.*` version of common.sql without a major disruption - because it woudl mean that all the providers that depend on common.sql will have to be upgraded as well (which will break the promise of more-or-less free upgrade and downgrade of providers independently of Airflow).

So as long as we get a breaking change in databricks provider, while keeping `common.sql` backwards-compatible, I am perfectly fine with having a breaking change here.

**Comment by potiuk** [comment]
(And I plan to add one ore two more ADRs for that  to explain it )

**Comment by bolkedebruin** [comment]
Sounds like a plan then! Thanks for your patience @Joffreybvn

**Review by potiuk — APPROVED** [review]

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (Joffreybvn) [inline airflow/providers/databricks/hooks/databricks_sql.py:198]: Mypy was not happy with `-> tuple | list[tuple] | list[list[tuple] | tuple]` as output - which is correct. I got error: "*Overloaded function signatures 1 and 2 overlap with incompatible return types*".

[Following mypy docs](https://mypy.readthedocs.io/en/stable/more_types.html#type-checking-the-variants), I added a `None` for it to be stop complaining. But that's not correct, this method cannot return None when a handler is provided...

Implementing definitive typing is out of the scope of this PR, and I admit mypy beat me on that case. Thus I let that for #36224 - but I'm whiling to help on that issue too of course.

File: airflow/providers/databricks/hooks/databricks_sql.py
Comment (Joffreybvn) [reply airflow/providers/databricks/hooks/databricks_sql.py]: I refactored `_make_common_data_structure` with the suggested typing, without None.

**Comment by Joffreybvn** [comment]
To be honest, I found quite interesting to be in controversial PR. Thus, totally fine for me :)

Current status:
- Refactored `_make_common_data_structure` with the suggested typing, without None.
- No more complex instance check, this function assume to have Row as input
- Little hack to make mypy happy [here](https://github.com/apache/airflow/pull/36205#discussion_r1434484284), but let's fix that definitively with @potiuk's issue #36224 ?

**Review by bolkedebruin — APPROVED** [review]
Much much better! I understand what this is about now. The follow up would make things even clearer so absolutely appreciated if you want to help out there.

**Comment by potiuk** [comment]
Hey @bolkedebruin -> next time when you merge PR - can you please make sure you keep "squash and merge" for PRs - this one has been merged without it and it has no (#36205) in the name - wnich made it disappear from our release notes generation.  I noticed it accidentally luckily.

**Comment by potiuk** [comment]
Screenshot 2023-12-23 at 01 49 34

**Comment by bolkedebruin** [comment]
Sorry @potiuk i did actually squash and merge but adjusted the commit message to cover what was done. That removed the ref. I'll be more careful

**Comment by potiuk** [comment]
No worries :). I caught it luckily

**Pull request merged**
